### DISCUSSION TITLE
problem with revalidateTag after initial deployment

### DISCUSSION THREAD

**Discussion by szymonhernik** [vercel/next.js#58287] [Help]
### Summary

**What I'm trying to accomplish**

I want to update the projects listed on the homepage with the newest data from CMS (Sanity) and achieve this using revalidateTag.

Set up: Next (14.0.1.), Sanity (CMS), Vercel (deployment). 

**Main problem**
The **revalidateTag** doesn't work. I'm using App router for the project and the problem appears when I fetch data in SSR function (that's where I also use the revalidateTag). 
It only worked (fetched the updates from CMS) after the first build on first refresh and then it stops fetching the newest updates from CMS and goes back to the data fetched during the first app build. 

My project repo: https://github.com/szymonhernik/debora-mora-portfolio_ts/tree/bug-revalidateTag

**Relevant chunks of code**

**Sanity webhook**
I have Sanity webhook for calling api/revalidate triggered on: "Create", "Update", and "Delete".

Simplified server function where I call the fetching function getProject:
```
import { getProjects } from '@root/sanity/sanity-utils'
import Link from 'next/link'

export default async function ProjectsGrid() {
  const projects = await getProjects()
  return (
          {projects.map((project) => (
            <Link href={`/projects/${project.slug}`} key={project._id}>
                <p>{project.name}</p>  
            </Link>
          ))}
  )
}
```

in `sanity/sanity-utils.ts` I have a SanityFetch and getProject functions:
 ```
import { Project } from '@root/types/Project'
import { createClient, groq } from 'next-sanity'
import clientConfig from './config/client-config'
import { revalidateSecret } from './sanity.api'
import type { QueryParams } from '@sanity/client'

const DEFAULT_PARAMS = {} as QueryParams
const DEFAULT_TAGS = [] as string[]

// Helper function to centralize the fetching logic including revalidation tags
export async function sanityFetch<QueryResponse>({
  query,
  params = DEFAULT_PARAMS,
  tags = DEFAULT_TAGS,
}: {
  query: string
  params?: QueryParams
  tags: string[]
}): Promise<QueryResponse> {
  const client = createClient(clientConfig).withConfig({
    useCdn: process.env.NODE_ENV === 'production' && !revalidateSecret, // Use CDN if in production and no revalidateSecret set
  })

  return client.fetch<QueryResponse>(query, params, {
    next: {
      tags,
    },
  })
}

export function getProjects() {
  const query = groq`*[_type == "project"]{
        _id,
        _createdAt,
        name,
        "slug":slug.current,
        "image": image.asset->url,
        url,
        content }`
  const tags = ['project', 'page']
  return sanityFetch<Project[]>({
    query,
    tags,
  })
}
```
and `app/api/revalidate/route.ts`
```
import { revalidateSecret } from '@root/sanity/sanity.api'
import { revalidateTag } from 'next/cache'
import { type NextRequest, NextResponse } from 'next/server'
import { parseBody } from 'next-sanity/webhook'

export async function POST(req: NextRequest) {
  try {
    const { isValidSignature, body } = await parseBody<{
      _type: string
      slug?: string | undefined
    }>(req, revalidateSecret)

    if (!isValidSignature) {
      const message = 'Invalid signature'
      return new Response(JSON.stringify({ message, isValidSignature, body }), { status: 401 })
    }

    if (!body?._type) {
      const message = 'Bad Request'
      return new Response(message, { status: 400 })
    }

    // If the `_type` is `page`, then all `client.fetch` calls with
    // `{next: {tags: ['page']}}` will be revalidated
    revalidateTag(body._type)
    if (body.slug) {
      revalidateTag(`${body._type}:${body.slug}`)
    }
    console.log(body)
    console.log(body)
    return NextResponse.json({
      status: 200,
      revalidated: true,
      now: Date.now(),
      body,
    })
  } catch (err: any) {
    console.error(err)
    return new Response(err.message, { status: 500 })
  }
}

```

**The logs/errors I am receiving**

**Vercel**
The logs in Vercel suggest success with the api revalidation that is called after revalidating (`[POST] /api/revalidate status=200`), however the GET calls don't fetch the updated data.

The `console.log(body)` of the tag body in Vercel 
`{ _type: 'project', slug: 'five-trees-change' }`

**What I've tried to solve the issue:**

I've tried following this template that works very well with the tag revalidation but it also included all the code responsible for it at the first build to github and vercel. It's from sanity.io team:  https://github.com/sanity-io/sanity-template-nextjs-app-router-personal-website

One things that somehow fixes the issue **temporarily** is purging cache in Vercel project dashboard. BUT, as soon as i push new code to github it goes back to the data from the first build. 

### Additional information

```js
perating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:39:46 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6020
Binaries:
  Node: 18.15.0
  npm: 9.5.0
  Yarn: N/A
  pnpm: 8.6.2
Relevant Packages:
  next: 14.0.1
  eslint-config-next: 14.0.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Example

https://github.com/szymonhernik/debora-mora-portfolio_ts

**Comment by sslowe** [comment]
Did you ever find a solution for this? Currently running into the same issue with Next + Sanity reverting to the build cache after first calling revalidateTag - only difference is my app is deployed via Amplify.
